scarb/src/core/registry/client/http.rs

```diff
@@ -0,0 +1,163 @@
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use fs4::tokio::AsyncFileExt;
+use futures::StreamExt;
+use reqwest::StatusCode;
+use tokio::fs::OpenOptions;
+use tokio::io;
+use tokio::io::BufWriter;
+use tokio::sync::OnceCell;
+use tracing::{debug, trace};
+
+use crate::core::registry::client::RegistryClient;
+use crate::core::registry::index::{IndexConfig, IndexRecords};
+use crate::core::{Config, Package, PackageId, PackageName, SourceId};
+use crate::flock::{FileLockGuard, Filesystem};
+
+// TODO(mkaput): Honour ETag and Last-Modified headers.
+// TODO(mkaput): Progressbar.
+// TODO(mkaput): Request timeout.
+
+/// Remote registry served by the HTTP-based registry API.
+pub struct HttpRegistryClient<'c> {
+    source_id: SourceId,
+    config: &'c Config,
+    cached_index_config: OnceCell<IndexConfig>,
+    dl_fs: Filesystem<'c>,
+}
+
+impl<'c> HttpRegistryClient<'c> {
+    pub fn new(source_id: SourceId, config: &'c Config) -> Result<Self> {
+        let dl_fs = config
+            .dirs()
+            .registry_dir()
+            .into_child("dl")
+            .into_child(source_id.ident());
+
+        Ok(Self {
+            source_id,
+            config,
+            cached_index_config: Default::default(),
+            dl_fs,
+        })
+    }
+
+    async fn index_config(&self) -> Result<&IndexConfig> {
+        // TODO(mkaput): Cache config locally, honouring ETag and Last-Modified headers.
+
+        async fn load(source_id: SourceId, config: &Config) -> Result<IndexConfig> {
+            let index_config_url = source_id
+                .url
+                .join(IndexConfig::WELL_KNOWN_PATH)
+                .expect("Registry config URL should always be valid.");
+            debug!("fetching registry config: {index_config_url}");
+
+            let index_config = config
+                .http()?
+                .get(index_config_url)
+                .send()
+                .await?
+                .error_for_status()?
```

Comment:
Should we have some retry policy for all this calls here?

Reply:
I don't think so. This is not a half-autonomous service, but just a one-off command line application. If something is off, we print the issue to the user, and they interactively decide whether to restart or no.

Reply:
Yes, but on the other hand, we make a bunch of them. That's why I wonder if we might have a problem if one download out of say 20 fails on unreliable network? 🤔

Reply:
With caching, the client will be idempotently dealing with this case and only pulling information it is missing so I think this would mitigate this issue?

Reply:
Also FYI: Cargo doesn't do retries and I don't think any other pkgmgr does